crates/ra_syntax/src/lib.rs

```diff
@@ -318,3 +318,92 @@ fn api_walkthrough() {
     }
     assert_eq!(exprs_cast, exprs_visit);
 }
+
+/// Generates `if let` chain to match provided `SyntaxNode` with `ast::*` nodes and to call
+/// corresponding function with node that matched. Collects syntax errors to the
+/// `Vec<SyntaxError>` provided.
+///
+/// Syntax:
+///
+/// ```notrust
+/// match_ast! {
+///     match <NODE_VAR> {
+///         ast::<NODE TYPE1> => <HANDLER FN1>, &mut <ERRORS VEC>,
```

Comment:
Or maybe use `ast::Literal(it) => validate_literal(it, &mut errors);`. That would also unify both macros. You can match the later using `:block`. Note that I used `;` as separator, because `:block` doesnt allow a `,` after it.

Reply:
I do not quite understand how can I match with the left-hand side of this line in the macro? If the `it` is used as an identifier, it must be declared somewhere first.

Reply:
Use `$p:ty($it:ident)`. I hope `(` is allowed to follow a ty though.

Reply:
Unfortunately, `(` is not allowed. Allowed tokens are: `{`, `[`, `=>`, `,`, `>`, `=`, `:`, `;`, `|`, `as` and `where`.

Reply:
It should be possible to work around this by using TT muncher, but it would be a mess.

Reply:
Another workaround would be to accept an ident for `p` and prepend the `ast::` part in the macro itself.